### PR TITLE
[navigation-next] feat: introduce new interface for group

### LINKED ISSUE #7061: Introduce `group` interfaces to OpenSearch Dashboards chrome service
**Is your feature request related to a problem? Please describe.**

`group` is a new concept to organize applications and categories, it will help users to focus on a group of features within OpenSearch Dashboards.

**Describe the solution you'd like**

Introduce two interfaces: addNavToGroup(set) and getGroupsMap$(get) into chrome_service so other plugins can register their existing or new features into specific group. 

**Describe alternatives you've considered**

**Additional context**

#7029

### PR DESCRIPTION
**Pull request by SuZhou-Joe**
### Description

This PR mainly introduces new interface to register and get useCases from chrome service, and it is a fundamental change for the new navigation.

### Issues Resolved

closes #7061 

## Screenshot

No UI changes.

## Testing the changes

As there is no place using the new interfaces introduced in this PR. You can refer to the test cases here:
src/core/public/chrome/chrome_service.test.ts

## Changelog

- feat: Introduce new interface for group

### Check List

- [x] All tests pass
  - [x] `yarn test:jest`
  - [x] `yarn test:jest_integration`
- [x] New functionality includes testing.
- [x] New functionality has been documented.
- [x] Update [CHANGELOG.md](./../CHANGELOG.md)
- [x] Commits are signed per the DCO using --signoff

### CODE REVIEW

File: src/core/public/chrome/chrome_service.tsx [hunk 1]
```diff
@@ -496,3 +534,19 @@ export interface InternalChromeStart extends ChromeStart {
...
   getHeaderComponent(): JSX.Element;
 }
+
+/** @public */
+export interface ChromeUseCase {
+  id: string;
+  title: string;
+  description: string;
+  order?: number;
+  icon?: string;
```
Comment (Hailong-am) [inline src/core/public/chrome/chrome_service.tsx:544]: should the icon type be `EuiIconType`?

File: src/core/public/chrome/chrome_service.tsx [hunk 2]
```diff
@@ -111,6 +111,9 @@ export class ChromeService {
   private readonly navLinks = new NavLinksService();
   private readonly recentlyAccessed = new RecentlyAccessedService();
   private readonly docTitle = new DocTitleService();
+  private readonly useCaseList$ = new BehaviorSubject<
+    Array<ChromeUseCase & { navLinks: ChromeRegistrationNavLink[] }>
+  >([]);
```
Comment (Hailong-am) [inline src/core/public/chrome/chrome_service.tsx:116]: could this be a map?

File: src/core/public/chrome/chrome_service.tsx [hunk 3]
```diff
@@ -158,6 +161,29 @@ export class ChromeService {
         }
         this.collapsibleNavHeaderRender = render;
       },
+      registerNavLink: (useCase: ChromeUseCase, navLink: ChromeRegistrationNavLink) => {
```
Comment (ruanyl) [inline src/core/public/chrome/chrome_service.tsx:164]: Since we're designing a public API, I feel the term `use case` is too specific because it's a product concept. Can we give it a more general name? something like `virtualDirectory`?

Comment (Hailong-am) [inline src/core/public/chrome/chrome_service.tsx:164]: registerUserCase? to pair with getUserCase$. Also can add an example here for better understanding

Comment (ruanyl) [inline src/core/public/chrome/chrome_service.tsx:164]: The function name `registerNavLink` doesn't sound appropriate, seems what we are doing here is register a relation between the app and the use case, but not register a specific nav link.

File: src/core/public/chrome/chrome_service.tsx [hunk 4]
```diff
@@ -158,6 +161,29 @@ export class ChromeService {
...
+        const currentUseCaseList = [...this.useCaseList$.getValue()];
+        const findItem = currentUseCaseList.find((item) => item.id === useCase.id);
```
Comment (ruanyl) [inline src/core/public/chrome/chrome_service.tsx:166]: can we use a meaningful name instead of`findItem`?

File: src/core/public/chrome/chrome_service.tsx [hunk 5]
```diff
@@ -496,3 +534,19 @@ export interface InternalChromeStart extends ChromeStart {
...
+  id: string;
+  title: string;
+  description: string;
+  order?: number;
+  icon?: string;
+  reserved?: boolean; // use cases like data administration / settings and setup is not configurable through workspace
+}
+
+/** @public */
+export interface ChromeRegistrationNavLink
```
Comment (ruanyl) [inline src/core/public/chrome/chrome_service.tsx:549]: It seems we only need `id` and `category` for now, shall we minimize this interface for the initial implementation?

File: src/core/public/chrome/chrome_service.tsx (hunk 2 above)
Comment (SuZhou-Joe) [reply src/core/public/chrome/chrome_service.tsx:116]: Good idea, internally it is more convenient to insert items into a map, and I am thinking make navLinks as a map as well. But I'd like to keep the output of `getUseCase$` as array.

File: src/core/public/chrome/chrome_service.tsx (hunk 3 above)
Comment (SuZhou-Joe) [reply src/core/public/chrome/chrome_service.tsx]: Changed to `addNavToGroup`, as for example, we add an example in the README of chrome_service, I suppose that should be good enough.

Comment (SuZhou-Joe) [reply src/core/public/chrome/chrome_service.tsx]: Yeah, changed to `group`.

Comment (SuZhou-Joe) [reply src/core/public/chrome/chrome_service.tsx]: Changed to `addNavToGroup`.

File: src/core/public/chrome/chrome_service.tsx (hunk 4 above)
Comment (SuZhou-Joe) [reply src/core/public/chrome/chrome_service.tsx]: Sure, changed the code here.

File: src/core/public/chrome/chrome_service.tsx (hunk 5 above)
Comment (SuZhou-Joe) [reply src/core/public/chrome/chrome_service.tsx]: Sure, let me make this interface simple for initial implementation

File: src/core/public/chrome/chrome_service.tsx [hunk 6]
```diff
@@ -158,6 +164,31 @@ export class ChromeService {
         }
         this.collapsibleNavHeaderRender = render;
       },
+      addNavToGroup: (navGroup: ChromeNavGroup, navLink: ChromeRegistrationNavLink) => {
+        // Construct a new groups map pointer.
+        const currentGroupsMap = { ...this.groupsMap$.getValue() };
+        const matchedGroup = currentGroupsMap[navGroup.id];
+        if (matchedGroup) {
+          const links = matchedGroup.navLinks;
+          const IfNavExist = !!links.find((link) => link.id === navLink.id);
```
Comment (wanglam) [inline src/core/public/chrome/chrome_service.tsx:173]: How about renaming to `isLinkExist`?

File: src/core/public/chrome/chrome_service.tsx [hunk 7]
```diff
@@ -158,6 +164,31 @@ export class ChromeService {
...
+        } else {
+          currentGroupsMap[navGroup.id] = {
+            ...navGroup,
+            navLinks: [navLink],
+          };
+        }
+
+        this.groupsMap$.next(currentGroupsMap);
+      },
+      getGroupsMap$: () => this.groupsMap$.pipe(takeUntil(this.stop$)),
```
Comment (Hailong-am) [inline src/core/public/chrome/chrome_service.tsx:191]: maybe `getNavGroupsMap`?

File: src/core/public/chrome/chrome_service.tsx (hunk 6 above)
Comment (SuZhou-Joe) [reply src/core/public/chrome/chrome_service.tsx:173]: Sure, updated

File: src/core/public/chrome/chrome_service.tsx [hunk 8]
```diff
@@ -496,3 +533,25 @@ export interface InternalChromeStart extends ChromeStart {
...
+}
+
+/** @public */
+export interface ChromeNavGroup {
+  id: string;
+  title: string;
+  description: string;
+  order?: number;
+  icon?: EuiIconType;
+  type: GROUP_TYPE; // group with type of GROUP_TYPE.SYSTEM_GROUP will always display before USE_CASE_GROUP
```
Comment (Hailong-am) [inline src/core/public/chrome/chrome_service.tsx:549]: can this accomplished by order if it only impacts the display order

**Review by Hailong-am — APPROVED** [review]

File: src/core/public/chrome/chrome_service.tsx [hunk 9]
```diff
@@ -158,6 +164,31 @@ export class ChromeService {
         }
         this.collapsibleNavHeaderRender = render;
       },
+      addNavToGroup: (navGroup: ChromeNavGroup, navLink: ChromeRegistrationNavLink) => {
```
Comment (AMoo-Miki) [inline src/core/public/chrome/chrome_service.tsx:167]: Since we are adding `...NavLink` items, this should at least be named `addNavLinkToGroup`.

However, the more I think of it, I find it uncomfortable that, as a consumer of this:
* I have to call it numerous times with a `navGroup` object that I make.
* Changes I might make to the `navGroup` will not be picked up.

A better way to mitigate this would be to make
```ts
addNavGroup: (navGroup: ChromeNavGroup, navLinks: [ChromeRegistrationNavLink], overwrite?: boolean)
```
... where the consumer adds a group and its links in one call.
1. We validate that the `navGroup.id` is unique when `overwrite` is false.
2. `navLink.id` is unique in the group.

File: src/core/public/chrome/chrome_service.tsx [hunk 10]
```diff
@@ -496,3 +533,25 @@ export interface InternalChromeStart extends ChromeStart {
    */
   getHeaderComponent(): JSX.Element;
 }
+
+export enum GROUP_TYPE {
+  SYSTEM_GROUP = 'SYSTEM_GROUP',
+  USE_CASE_GROUP = 'USE_CASE_GROUP',
+}
```
Comment (AMoo-Miki) [inline src/core/public/chrome/chrome_service.tsx:540]:
```suggestion
export enum NavGroupType {
  SYSTEM = 'system',
}
```
... and make `type` optional below.

Note that the `enum` name is in Pascal-case, the name of the the actual constant is uppercase, with no suffix tying it to the `enum` name, and the value is lowercase.

File: src/core/public/chrome/chrome_service.tsx (hunk 8 above)
Comment (AMoo-Miki) [reply src/core/public/chrome/chrome_service.tsx:549]: I agree with @Hailong-am; if the purpose of `type` is to display something higher, we could reserve `order` of negative for OSD et al. and positives for others.

File: src/core/public/chrome/chrome_service.tsx (hunk 10 above)
Comment (SuZhou-Joe) [reply src/core/public/chrome/chrome_service.tsx:540]: Thanks.

File: src/core/public/chrome/chrome_service.tsx (hunk 8 above)
Comment (SuZhou-Joe) [reply src/core/public/chrome/chrome_service.tsx:549]: My fault, not only the order. It indicates if the group can be picked up as a use case in workspace creation page. Let me add more comment on it.

File: src/core/public/chrome/chrome_service.tsx (hunk 9 above)
Comment (SuZhou-Joe) [reply src/core/public/chrome/chrome_service.tsx:167]: Hi @AMoo-Miki , I get your point but imaging that a group consists of applications from various plugins from various repos, there is not a centralized place to `addNavGroup`.

But I do think making addNavLinksToGroup to accept an array of navLinks makes sense. Will adjust that.

Comment (AMoo-Miki) [reply src/core/public/chrome/chrome_service.tsx]: Sadly it is not a 2-way door: being a public API, we will only be able to change it if we make a major version bump.

> imaging that there is a group consists of applications from various plugins in various repos, there is not a centralized place to addNavGroup

This makes the problem a bit more interesting.

```ts
private readonly navGroupsMap$ = new BehaviorSubject<Record<string, NavGroupItemInMap>>({
  '': { id: '', title: '', description: '', navLinks: [] }
});

addNavGroup(id: string, Omit<ChromeNavGroup, 'id'>, failIfExists?: boolean = false) {
  if (!id?.trim?.()) throw 'NavGroup `id` is expected to be a non-blank string';
  const groupsMap = { ...this.navGroupsMap$.getValue() };
  if (id in groupsMap) {
    if (failIfExists) throw 'Duplicate NavGroup `id` found';
    return false;
  }

  groupsMap[id] = {
    ...navGroup,
    id,
    navLinks,
  };
  this.navGroupsMap$.next(groupsMap);
}

addNavLinks(links: ChromeRegistrationNavLink[] | ChromeRegistrationNavLink, groupId?: string) {
  /// still working on this comment
}
```

File: src/core/public/chrome/README.md [hunk 11]
```diff
@@ -112,6 +112,31 @@ Gets an Observable of the array of recently accessed history :-
...
+
+`addNavLinksToGroup: (group: ChromeNavGroup, navLinks: ChromeRegistrationNavLink[]) => void;`
+
+Gets an Observable of the array of registered groups :-
+
+`getNavGroupsMap$: Observable<Record<string, NavGroupItemInMap>>`
+##### Register a new group with a navLink
+
+  ```ts
+    chrome.addNavLinksToGroup(
```
Comment (AMoo-Miki) [inline src/core/public/chrome/README.md:128]: This doesn't follow the pattern set in `chrome` where each service has its own accessor property.

**Comment by AMoo-Miki** [comment]
Are we deprecating the existing `NavLinksService`. We cannot have two methods doing the same thing simultaneously without marking one as deprecated.

PS, this doesn't mean that we are removing it right away; we are just warning people from using the old method.

File: src/core/public/chrome/README.md (hunk 11 above)
Comment (SuZhou-Joe) [reply src/core/public/chrome/README.md:128]: I see, will adjust the doc.

**Comment by SuZhou-Joe** [comment]
@AMoo-Miki No, we are not deprecating the NavLinksService, the service is a converter to transform applications into NavLinks and the navGroup is just a concept to group the NavLinks generated by NavLinksService. I'd more like to regard navGroup as a downstream of NavLinksService.

File: src/core/public/chrome/README.md (hunk 11 above)
Comment (SuZhou-Joe) [reply src/core/public/chrome/README.md]: Changed, move the example to the comment of interface.

File: src/core/public/chrome/nav_group/nav_group_service.ts [hunk 12]
```diff
@@ -0,0 +1,136 @@
...
+  }
+  private getSortedNavGroupsMap() {
+    return combineLatest([this.navGroupsMap$, this.navLinks$])
+      .pipe(takeUntil(this.stop$))
+      .pipe(
+        map(([navGroupsMap, navLinks]) => {
+          return Object.keys(navGroupsMap).reduce((sortedNavGroupsMap, navGroupId) => {
+            const navGroup = navGroupsMap[navGroupId];
+            const sortedNavLinks = getOrderedLinksOrCategories(
+              fullfillRegistrationLinksToChromeNavLinks(navGroup.navLinks, navLinks)
```
Comment (Hailong-am) [inline src/core/public/chrome/nav_group/nav_group_service.ts:88]: typo `fulfillRegistrationLinksToChromeNavLinks

File: src/core/utils/default_app_categories.ts [hunk 13]
```diff
@@ -73,4 +73,25 @@ export const DEFAULT_APP_CATEGORIES: Record<string, AppCategory> = Object.freeze
     order: 5000,
     euiIconType: 'managementApp',
   },
+  investigate: {
```
Comment (wanglam) [inline src/core/utils/default_app_categories.ts:76]: Are these new categories added by intentionally? Seems not related the new interface.

Comment (SuZhou-Joe) [reply src/core/utils/default_app_categories.ts:76]: Yes, is is added since these categories are default categories for core applications.

**Review by wanglam — APPROVED** [review]

File: src/core/public/chrome/nav_group/nav_group_service.ts (hunk 12 above)
Comment (SuZhou-Joe) [reply src/core/public/chrome/nav_group/nav_group_service.ts]: Thanks for catching that, updated.

**Review by seraphjiang — APPROVED** [review]

File: src/core/public/chrome/nav_group/nav_group_service.ts [hunk 14]
```diff
@@ -0,0 +1,136 @@
...
+    } else {
+      currentGroupsMap[navGroup.id] = {
+        ...navGroup,
+        navLinks: [navLink],
+      };
+    }
+
+    return currentGroupsMap;
+  }
+  private getSortedNavGroupsMap() {
```
Comment (ruanyl) [inline src/core/public/chrome/nav_group/nav_group_service.ts:80]: Shall we name it as `getSortedNavGroupsMap$()` since it returns observable?

File: src/core/public/chrome/utils.ts [hunk 15]
```diff
@@ -0,0 +1,150 @@
...
+}
+
+export function fulfillRegistrationLinksToChromeNavLinks(
+  registerNavLinks: ChromeRegistrationNavLink[],
+  navLinks: ChromeNavLink[]
+): Array<ChromeNavLink & ChromeRegistrationNavLink> {
+  const allExistingNavLinkId = navLinks.map((link) => link.id);
+  return (
+    registerNavLinks
+      ?.filter((navLink) => allExistingNavLinkId.includes(navLink.id))
```
Comment (ruanyl) [inline src/core/public/chrome/utils.ts:40]: Could `registerNavLinks` be undefined? Maybe optional chain is unnecessary, or `registerNavLinks` should be defined as optional.

File: src/core/public/chrome/utils.ts [hunk 16]
```diff
@@ -0,0 +1,150 @@
...
+  const allCategories = {} as Record<string, AppCategory | undefined>;
+
+  for (const [key, value] of Object.entries(allCategorizedLinks)) {
+    allCategories[key] = value[0].category;
+  }
+
+  return allCategories;
+}
+
+export function fulfillRegistrationLinksToChromeNavLinks(
```
Comment (ruanyl) [inline src/core/public/chrome/utils.ts:33]: Would you mind to leave a comment to describe this function?

File: src/core/public/chrome/utils.ts [hunk 17]
```diff
@@ -0,0 +1,150 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+import { AppCategory } from 'opensearch-dashboards/public';
+import { groupBy, sortBy } from 'lodash';
```
Comment (ruanyl) [inline src/core/public/chrome/utils.ts:7]: Shall we avoid using `lodash`? As I have seen multiple issues about getting rid of `lodash`.

File: src/core/public/chrome/utils.ts [hunk 18]
```diff
@@ -0,0 +1,150 @@
...
+  const navLinksUnderParentId = navLinksGoupedByParentNavLink[navLink.id];
+
+  if (navLinksUnderParentId) {
+    return {
+      itemType: LinkItemType.PARENT_LINK,
+      link: navLink,
+      links: getOrderedLinks(navLinksUnderParentId || []).map((navLinkUnderParentId) =>
+        generateItemTypeByLink(navLinkUnderParentId, navLinksGoupedByParentNavLink)
+      ),
+      order: navLink?.order,
```
Comment (ruanyl) [inline src/core/public/chrome/utils.ts:81]: Nit: this optional chain seems unnecessary

File: src/core/public/chrome/utils.ts [hunk 19]
```diff
@@ -0,0 +1,150 @@
...
+
+    return acc;
+  }, [] as ChromeNavLink[]);
+}
+
+export const generateItemTypeByLink = (
+  navLink: ChromeNavLink & ChromeRegistrationNavLink,
+  navLinksGoupedByParentNavLink: Record<string, ChromeNavLink[]>
+): LinkItem => {
+  const navLinksUnderParentId = navLinksGoupedByParentNavLink[navLink.id];
```
Comment (ruanyl) [inline src/core/public/chrome/utils.ts:72]: Typo:
```suggestion
  const navLinksUnderParentId = navLinksGroupedByParentNavLink[navLink.id];
```

File: src/core/types/nav_group.ts [hunk 20]
```diff
@@ -0,0 +1,28 @@
...
+  description: string;
+  order?: number;
+  icon?: EuiIconType;
+
+  /**
+   * Groups with type of NavGroupType.SYSTEM will:
+   * 1. Always display before USE_CASE_GROUP.
+   * 2. Not be pickable within the workspace creation page.
+   *
+   * @default undefined indicates it is of type useCase
```
Comment (ruanyl) [inline src/core/types/nav_group.ts:25]: Should this comment be adjusted a bit? Maybe the comment should be not in core, but at where the `type` is used.

File: src/core/public/chrome/nav_group/nav_group_service.ts [hunk 21]
```diff
@@ -0,0 +1,136 @@
...
+            return sortedNavGroupsMap;
+          }, {} as Record<string, NavGroupItemInMap>);
+        })
+      );
+  }
+  setup({ uiSettings }: { uiSettings: IUiSettingsClient }): ChromeNavGroupServiceSetupContract {
+    this.navGroupEnabledUiSettingsSubscription = uiSettings
+      .get$('home:useNewHomePage', false)
+      .subscribe((value) => {
+        this.navGroupEnabled = value;
```
[the remark this answers is not in the document]
Comment (wanglam) [reply src/core/public/chrome/nav_group/nav_group_service.ts:103]: Shall we add `requiresPageReload: true,` in `src/plugins/home/server/ui_settings.ts`? Then the page will show a toast `One or more settings require you to reload the page to take effect.`. The `navGroupEnabled` is a private property in `NavGroupService`, I think it can't be used outside directly.

[the remark this answers is not in the document]
Comment (SuZhou-Joe) [reply src/core/public/chrome/nav_group/nav_group_service.ts:103]: Agree, let's add requiresPageReload to the setting.

File: src/core/public/chrome/nav_group/nav_group_service.ts (hunk 14 above)
Comment (SuZhou-Joe) [reply src/core/public/chrome/nav_group/nav_group_service.ts]: Sure, updated

File: src/core/public/chrome/utils.ts (hunk 16 above)
Comment (SuZhou-Joe) [reply src/core/public/chrome/utils.ts:58]: Sure, done for that.

File: src/core/public/chrome/utils.ts (hunk 15 above)
Comment (SuZhou-Joe) [reply src/core/public/chrome/utils.ts]: You are right, remove the unnecessary optional chain.

File: src/core/public/chrome/utils.ts (hunk 17 above)
Comment (SuZhou-Joe) [reply src/core/public/chrome/utils.ts]: Sure, updated to use native approach to group and sort.

File: src/core/public/chrome/utils.ts (hunk 19 above)
Comment (SuZhou-Joe) [reply src/core/public/chrome/utils.ts]: Thanks for catching this, updated.

File: src/core/types/nav_group.ts (hunk 20 above)
Comment (SuZhou-Joe) [reply src/core/types/nav_group.ts]: Sure, move the comment to the enum NavGroupType, and the place which will use this type is in another PR and still waiting for finalized design.

File: src/core/public/chrome/utils.ts (hunk 18 above)
Comment (SuZhou-Joe) [reply src/core/public/chrome/utils.ts]: Sure, remove that.

File: src/core/public/chrome/nav_group/nav_group_service.ts [hunk 22]
```diff
@@ -0,0 +1,136 @@
...
+  private navGroupEnabledUiSettingsSubscription: Subscription | undefined;
+  private addNavLinkToGroup(
+    currentGroupsMap: Record<string, NavGroupItemInMap>,
+    navGroup: ChromeNavGroup,
+    navLink: ChromeRegistrationNavLink
+  ) {
+    const matchedGroup = currentGroupsMap[navGroup.id];
+    if (matchedGroup) {
+      const links = matchedGroup.navLinks;
+      const isLinkExistInGroup = !!links.find((link) => link.id === navLink.id);
```
[the remark this answers is not in the document]
Comment (SuZhou-Joe) [reply src/core/public/chrome/nav_group/nav_group_service.ts]: Sure, done for that.

**Review by ruanyl — APPROVED** [review]

**Comment by seraphjiang** [comment]
@BionIT @ZilongX would you help to review and merge?

**Review by BionIT — APPROVED** [review]

**Comment by zhyuanqi** [comment]
https://github.com/opensearch-project/OpenSearch-Dashboards/pull/7136 Remove failed backport as it has been manually backported.
